webSockets/client/client.go

```diff
@@ -32,138 +31,121 @@ type client struct {
+		return data.ErrZeroValueRetryDuration
+	}
+	return nil
+}
 
-	_ = c.sender.AddConnection(c.wsConn)
-
-	timer := time.NewTimer(c.retryDuration)
-	defer timer.Stop()
-	for {
-		err := c.sender.Send(args.Payload)
-		if err == nil {
-			return nil
-		}
+// Start will start the websockets client (will initialize a connection with the ws server)
+func (c *client) Start() {
+	go func() {
+		timer := time.NewTimer(c.retryDuration)
+		defer timer.Stop()
 
-		_, isConnectionClosed := err.(*websocket.CloseError)
-		shouldOpenNewConnection := isConnectionClosed ||
-			strings.Contains(err.Error(), data.ClosedConnectionMessage) ||
-			strings.Contains(err.Error(), data.CloseSent)
-		if shouldOpenNewConnection {
-			// open a new connection
-			c.log.Warn("clientSender: the previous connection was closed -> trying to open a new connection")
-			c.wsConn = connection.NewWSConnClient()
-			closed = c.openConnection()
-			if closed {
-				return nil
+		for {
+			err := c.wsConn.OpenConnection(c.url)
+			if err != nil && !errors.Is(err, data.ErrConnectionAlreadyOpened) {
+				c.log.Warn(fmt.Sprintf("c.openConnection(), retrying in %v...", c.retryDuration), "error", err)
 			}
-			// the old connection will be rewritten in the internal map
-			_ = c.sender.AddConnection(c.wsConn)
-			continue
-		}
-
-		c.log.Warn(fmt.Sprintf("client.writeMessage: connection problem retrying in %v", c.retryDuration), "error", err)
 
-		timer.Reset(c.retryDuration)
+			timer.Reset(c.retryDuration)
 
-		select {
-		case <-timer.C:
-		case <-c.safeCloser.ChanClose():
-			return nil
+			select {
+			case <-timer.C:
+			case <-c.safeCloser.ChanClose():
+				return
+			}
 		}
-	}
+	}()
+}
+
+// Send will send the provided payload from args
+func (c *client) Send(args data.WsSendArgs) error {
+	return c.sender.Send(args.Payload)
 }
 
-// RegisterPayloadHandler register the payload handler
-func (c *client) RegisterPayloadHandler(handler webSockets.PayloadHandler) {
-	c.receiver.SetPayloadHandler(handler)
+// SetPayloadHandler set the payload handler
+func (c *client) SetPayloadHandler(handler webSockets.PayloadHandler) error {
+	return c.receiver.SetPayloadHandler(handler)
 }
 
 // Listen will listen from messages
 func (c *client) Listen() {
 	timer := time.NewTimer(c.retryDuration)
 	defer timer.Stop()
 
-	closed := false
-	for !closed {
-		c.openConnection()
-
-		closed = c.receiver.Listen(c.wsConn)
-
-		timer.Reset(c.retryDuration)
-
-		select {
-		case <-timer.C:
-		case <-c.safeCloser.ChanClose():
-			return
-		}
-	}
-}
-
-func (c *client) openConnection() (closed bool) {
-	timer := time.NewTimer(c.retryDuration)
-	defer timer.Stop()
-
 	for {
-		err := c.wsConn.OpenConnection(c.url)
-		if err == nil || errors.Is(err, data.ErrConnectionAlreadyOpened) {
-			return
-		} else {
-			c.log.Warn(fmt.Sprintf("c.openConnection(), retrying in %v...", c.retryDuration), "error", err)
-		}
+		_ = c.receiver.Listen(c.wsConn)
 
 		timer.Reset(c.retryDuration)
```

Comment:
Listen calls should be done in a very rapid manner. I would completely remove the timer here and simplify the function as follows:
```GO
func (c *client) Listen() {
    for {
        _ = c.receiver.Listen(c.wsConn)
        
        select {
        case <-c.safeCloser.ChanClose():
	    return
        }
    }
}

```

Reply:
refactored.